File: backend/app/routers/lines.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from typing import List, Optional
from pydantic import BaseModel

from app import schemas, crud
from app.database import get_db

router = APIRouter(
    prefix="/lines",
    tags=["lines"]
)


@router.post("", response_model=schemas.Line, status_code=201)
def create_line(line: schemas.LineCreate, db: Session = Depends(get_db)):
    # Verify team exists
    team = crud.get_team(db, line.team_id)
    if not team:
        raise HTTPException(status_code=404, detail="Team not found")

    # Verify all players exist and belong to the team
    if line.player_ids:
        players = crud.get_players_by_team(db, line.team_id)
        player_ids_set = {p.id for p in players}
        invalid_players = set(line.player_ids) - player_ids_set
        if invalid_players:
            raise HTTPException(
                status_code=400,
                detail=f"Players {invalid_players} not found in team {line.team_id}"
            )

    try:
        return crud.create_line(db, line)
    except IntegrityError:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=f"Line with name '{line.name}' already exists for this team"
        )


@router.get("", response_model=List[schemas.LineWithPlayers])
def list_lines(team_id: Optional[int] = None, skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return crud.get_lines(db, team_id=team_id, skip=skip, limit=limit)


@router.get("/{line_id}", response_model=schemas.LineWithPlayers)
def get_line(line_id: int, db: Session = Depends(get_db)):
    line = crud.get_line(db, line_id)
    if not line:
        raise HTTPException(status_code=404, detail="Line not found")
    return line


@router.put("/{line_id}", response_model=schemas.Line)
def update_line(line_id: int, line_update: schemas.LineUpdate, db: Session = Depends(get_db)):
    try:
        line = crud.update_line(db, line_id, line_update)
        if not line:
            raise HTTPException(status_code=404, detail="Line not found")
        return line
    except IntegrityError:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=f"Line with name '{line_update.name}' already exists for this team"
        )


@router.delete("/{line_id}", status_code=204)
def delete_line(line_id: int, db: Session = Depends(get_db)):
    success = crud.delete_line(db, line_id)
    if not success:
        raise HTTPException(status_code=404, detail="Line not found")


# Nested endpoints for line resources
@router.get("/{line_id}/players", response_model=List[schemas.Player])
def list_line_players(line_id: int, db: Session = Depends(get_db)):
    """Get all players in this line"""
    line = crud.get_line(db, line_id)
    if not line:
        raise HTTPException(status_code=404, detail="Line not found")
    return crud.get_line_players(db, line_id)


class PlayerIdsRequest(BaseModel):
    player_ids: List[int]


@router.post("/{line_id}/players", response_model=schemas.LineWithPlayers)
def add_players_to_line(line_id: int, request: PlayerIdsRequest, db: Session = Depends(get_db)):
    """Add players to line"""
    line = crud.get_line(db, line_id)
    if not line:
        raise HTTPException(status_code=404, detail="Line not found")

    # Verify all players belong to the same team
    if request.player_ids:
        players = crud.get_players_by_team(db, line.team_id)
        player_ids_set = {p.id for p in players}
        invalid_players = set(request.player_ids) - player_ids_set
        if invalid_players:
            raise HTTPException(
                status_code=400,
                detail=f"Players {invalid_players} not found in team {line.team_id}"
            )

    return crud.add_players_to_line(db, line_id, request.player_ids)


@router.delete("/{line_id}/players", response_model=schemas.LineWithPlayers)
def remove_players_from_line(line_id: int, request: PlayerIdsRequest, db: Session = Depends(get_db)):
    """Remove players from line"""
    line = crud.get_line(db, line_id)
    if not line:
        raise HTTPException(status_code=404, detail="Line not found")

    return crud.remove_players_from_line(db, line_id, request.player_ids)

File: frontend/src/services/lines.ts
import { apiClient } from "./api";
import type {
  Line,
  LineCreate,
  LineUpdate,
  LineWithPlayers,
  Player,
  PlayerIdsRequest,
} from "../types";

// Create a new line
export const createLine = async (data: LineCreate): Promise<Line> => {
  const response = await apiClient.post<Line>("/lines", data);
  return response.data;
};

// Get all lines (optionally filtered by team)
export const getLines = async (teamId?: number): Promise<LineWithPlayers[]> => {
  const params = teamId ? { team_id: teamId } : {};
  const response = await apiClient.get<LineWithPlayers[]>("/lines", { params });
  return response.data;
};

// Get line by ID with players
export const getLine = async (lineId: number): Promise<LineWithPlayers> => {
  const response = await apiClient.get<LineWithPlayers>(`/lines/${lineId}`);
  return response.data;
};

// Update line
export const updateLine = async (
  lineId: number,
  data: LineUpdate
): Promise<Line> => {
  const response = await apiClient.put<Line>(`/lines/${lineId}`, data);
  return response.data;
};

// Delete line
export const deleteLine = async (lineId: number): Promise<void> => {
  await apiClient.delete(`/lines/${lineId}`);
};

// Get line players
export const getLinePlayers = async (lineId: number): Promise<Player[]> => {
  const response = await apiClient.get<Player[]>(`/lines/${lineId}/players`);
  return response.data;
};

// Add players to line
export const addPlayersToLine = async (
  lineId: number,
  playerIds: number[]
): Promise<LineWithPlayers> => {
  const response = await apiClient.post<LineWithPlayers>(
    `/lines/${lineId}/players`,
    { player_ids: playerIds } as PlayerIdsRequest
  );
  return response.data;
};

// Remove players from line
export const removePlayersFromLine = async (
  lineId: number,
  playerIds: number[]
): Promise<LineWithPlayers> => {
  const response = await apiClient.delete<LineWithPlayers>(
    `/lines/${lineId}/players`,
    { data: { player_ids: playerIds } as PlayerIdsRequest }
  );
  return response.data;
};
